ci: Delete VMs older than 23 hours

nbyavuz opened the pull request:
There could be some leftover VMs. Clear these VMs because of their costs and to get back the space their disks occupy.

It is set to get VMs older than 23 hours instead of 1 day because cron job runs daily, so there could be race conditions.

anarazel commented on gcp_delete_old_images_and_vms.py:
```diff
@@ -0,0 +1,102 @@
+        print('Deleting VM =', vm['name'])
+        delete_cmd = base_cmd + ['delete', vm['name'],
+                                 '--zone', vm['zone'], '--quiet']
+        subprocess.run(delete_cmd, check=True)
```
I think we also need to delete old disks for those VMs?

nbyavuz replied:
This command actually deletes VMs and their disks.

anarazel replied:
I think that's only true for disks explicitly owned by VMs - but the way packer creates disk they are not set up that way (because packer needs to access it after shutting the VM down).

nbyavuz replied:
Added a explicit way to deleting remaining disks older than 23 hours

anarazel commented on gcp_delete_old_images_and_vms.py:
```diff
@@ -0,0 +1,102 @@
+        sys.exit(1)
+
+    delete_vms()
+    # delete_families()
```
what's that about?

nbyavuz replied:
My bad, sorry.

anarazel commented on gcp_delete_old_images_and_instances.py:
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
```
Could you put the rename in a separate commit? Unfortunately it makes the diff hard to read...

anarazel commented on gcp_delete_old_images_and_instances.py:
```diff
@@ -0,0 +1,104 @@
+
+# Delete images older than 2 weeks, except if an image is the newest image in
+# an image family
+def delete_families():
```
This does *not* delete families. It deletes *older* images in families, but never the last image in a family.

anarazel commented on gcp_delete_old_images_and_instances.py:
```diff
@@ -0,0 +1,104 @@
+
+
+# Delete objects older than 23 hour
+def delete_objects(type: str):
```
I don't think "objects" is a good name - that'd include images, which are handled separately.  Maybe delete_leftover_running()?

anarazel commented on gcp_delete_old_images_and_instances.py:
```diff
@@ -0,0 +1,104 @@
+        sys.exit(1)
+
+    delete_objects('instances')
+    delete_objects('disks')
```
I think we should delete the instances / disks in a separate script. We only run the cleanup task if everything succeeded, but we are more likely to accumulate leftover things if there are failures. 

Also, if there are leftover objects, we might fail running due to resource constraints. So we should remove leftover things *before* we run packer.

nbyavuz commented:
I didn't revert changes on 'gcp_delete_old_images.py' file, they are mostly styling changes(using pycodestyle). Should I revert these changes too?

anarazel commented on .cirrus.yml:
```diff
@@ -151,6 +167,8 @@ task:
       matrix:
         - name: openbsd-vanilla
+          depends_on:
+           - cleanup-leftover-running
```
Can't we do this depend outside of the matrix, instead of repeating it for each "arm"?

nbyavuz replied:
If it is implemented like you said, 'depends_on at outside of matrix' overwrites open/netbsd-postgres task's depends_on. They starts to depend 'cleanup-leftover-running' instead of their vanilla tasks.

anarazel commented on gcp_delete_leftover_running.py:
```diff
@@ -0,0 +1,52 @@
+        print(f"no {type} to delete")
+
+    for elem in delete_list:
+        print(f'Deleting {type} =', elem['name'])
```
If you're using f-strings, why not also use them for `elem['name']`?

anarazel reviewed (commented):
Could you split the change to delete-old-images into a separate PR or at least separate commit?

anarazel reviewed (approved):
Looks good to me. Have you been able to test that this actually works as-is?

nbyavuz commented:
> Looks good to me. Have you been able to test that this actually works as-is?

Yes, I set time to 1 second and run it locally:

```sh
$ python3 gcp_delete_leftover_running.py 
instances to be deleted:  instance-1, instance-2
Deleting instances = $instance-1
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/instances/instance-1].
Deleting instances = $instance-2
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/instances/instance-2].
disks to be deleted:  disk-1
Deleting disks = $disk-1
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/disks/disk-1].
```

Just pushed a small fix for removing '$' sign while printing.

anarazel commented:
> Yes, I set time to 1 second and run it locally:

Cool, then let's merge this once the tests have passed.
